### PR TITLE
[WIP]Feature/search functionality

### LINKED ISSUE #1002: Adding Global Search Functionality
### What would you like to be added?

Listing Everything that has to be done here and also discussing the implementation.

Design :- https://www.figma.com/file/M3BpWO13WJVp7YivScL6K9/Dashboard?node-id=86%3A107
Search bar would be present on the navbar.

1)  **Auto Completion Search Results** :- Whenever User types in the search bar he would be shown possible search results from tables players, patrons and guilds as shown in design.

 Debouncing:- To prevent multiple queries to the database on every keystroke.
 Use react-select for building the UI instead of Input.

  Displaying Search Suggestions:- 
    How to make query faster as it would be searching across multiple tables ?
     Using #990 would help but would it give desirable performance  and not very slow? 
     Implementation fuzzy search:- https://hasura.io/docs/latest/graphql/core/databases/postgres/schema/custom-functions.html#example-fuzzy-match-search-functions
    
2)  **New Search Page**:- Which would display all the results and filter by Players, Patrons and Guilds.

   Managing the state of filters and displaying different components depending on what is active like Guilds or players and
   implementing infinite lazy loading.
   Search Query to search guild by name and indexing that table.

### Why is this needed?

To search from any page across different data.

### PR DESCRIPTION
**Pull request by Udit-takkar**
closes #1002 

To Improve

Implementation
- Create index and SQL functions on hasura console
- Created new pages for displaying results
- Used AsyncSelect component for search bar

## Assets

https://user-images.githubusercontent.com/53316345/152695045-f3028117-9bb9-4a1f-9a3a-d493a0c20d55.mov

### CODE REVIEW

**Comment by Udit-takkar** [comment]
Have to replace these Link href.

**Comment by Udit-takkar** [comment]
@vidvidvid  
I am stuck at few things.  Currently busy with few things. can you continue this feature?

Search Icon :- When i add this search icon in Async Select Component the input do not work properly. 

```
  Control: ({ children, ...rest }) => (
               <components.Control {...rest}>
                 <BoxedNextImage
                   src={SearchIcon}
                   alt="search"
                   height={5}
                   width={5}
                   ml="2"
                 />
                 {children}
               </components.Control>
            ```

 Patrons  & Guilds :- 

1) Would have to write a query for fetching Patrons by input text which would consist of executing multiple queries.
2) Infinite Lazy Loading is not set up for Guilds and Patrons Page.

I don't know sometimes the search input behaves wierdly becauses it caches the results

**Comment by vidvidvid** [comment]
> @vidvidvid I am stuck at few things
> 
> Search Icon :- When i add this search icon in Async Select Component the input does not work properly it do not take more than one character input.
> 
> ```
>   Control: ({ children, ...rest }) => (
>                <components.Control {...rest}>
>                  <BoxedNextImage
>                    src={SearchIcon}
>                    alt="search"
>                    height={5}
>                    width={5}
>                    ml="2"
>                  />
>                  {children}
>                </components.Control>
> ```

It looks to me like each time you press a key, this causes the input to rerender. Unfortunately I don't know how to solve this. :c

> Patrons & Guilds :-
> 
> 1. Would have to write a query for fetching Patrons by input text which would consist of executing multiple queries.
> 2. Infinite Lazy Loading is not set up for Guilds and Patrons Page.

If you feel like you can do that and it's required for you to finish this PR, go for it.

> I don't know sometimes the search input behaves wierdly becauses it caches the results

Can you give me and example of weird behaviour?

**Comment by vidvidvid** [comment]
Otherwise the preview of your PR looks great! Good job! :)

**Comment by Udit-takkar** [comment]
So Async Select is not able to properly invalidate cache. so when i type in some input the results are often not same as received from API response and sometimes no request is triggered and some previous results are displayed.

**Comment by vidvidvid** [comment]
> So Async Select is not able to properly invalidate cache. so when i type in some input the results are often not same as received from API response and sometimes no request is triggered and some previous results are displayed.

I know you did a lot of work already, but at this point I would consider writing your own component. It shouldn't be too hard and you would have more control over what and how you wanna display stuff - an input and a container with all the results.

You could also try out different libraries or contacting the developers of `react-select`.

File: packages/web/components/MegaMenu.tsx [hunk 1]
```diff
@@ -312,31 +318,194 @@ const DesktopNavLinks = () => {
...
+  handleClick: () => void;
+  name: string;
+  imgSrc?: string;
+  text: string;
+}) => (
+  <Flex align="center" onClick={handleClick}>
+    <Avatar name={name} src={imgSrc} w="20px" h="20px" />
+    <Text ml="2">{text}</Text>
+  </Flex>
+);
```
Comment (dysbulic) [inline packages/web/components/MegaMenu.tsx:375]:
```suggestion
const MenuOption = ({
  onClick,
  name,
  image: src,
  option,
}: {
  onClick: () => void;
  name: string;
  image?: string;
  option: string;
}) => (
  <Flex align="center" {...{ onClick }}>
    {image && <Avatar {...{ name, src }} w={5} h={5} />}
    <Text ml={2}>{option}</Text>
  </Flex>
);
```

File: packages/web/components/MegaMenu.tsx [hunk 2]
```diff
@@ -312,31 +318,194 @@ const DesktopNavLinks = () => {
...
 
+const Search = () => {
+  const [inputValue, setValue] = React.useState<string>('');
```
Comment (dysbulic) [inline packages/web/components/MegaMenu.tsx:378]:
```suggestion
  const [search, setSearch] = useState<string>();
```

File: packages/web/components/MegaMenu.tsx [hunk 3]
```diff
@@ -312,31 +318,194 @@ const DesktopNavLinks = () => {
...
+  const realtimeInput = React.useRef('');
+  const router = useRouter();
+
+  const debounced = useDebouncedCallback(
+    async () => {
+      const { players } = await getPlayersByText(`%${inputValue}%`);
+      const { guilds } = await getGuildsByText(`%${inputValue}%`);
```
Comment (dysbulic) [inline packages/web/components/MegaMenu.tsx:385]: The enclosing of the search text with `%`s should happen in the method calling the GraphQL query (i.e. *within* `searchPlayers`).

You just pass in the search string to these methods:

```suggestion
      const { players } = await searchPlayers(search);
      const { guilds } = await searchGuilds(search);
```

File: packages/web/graphql/getPlayers.ts [hunk 4]
```diff
@@ -194,3 +197,33 @@ export const getPlayerFilters = async (client: Client = defaultClient) => {
 
   return data;
 };
+
+// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+gql`
```
Comment (dysbulic) [inline packages/web/graphql/getPlayers.ts:202]:
```suggestion
/* GraphQL */ `
```

**Review by dysbulic — CHANGES_REQUESTED** [review]
@Udit-takkar, I started a thorough review, but that takes a while, and I'm working on a pull request of my own.

I've identified two major changes that need to happen:

* Remove the custom SQL functions and do the queries using the regular GraphQL endpoint
* Put `Search` in its own file and import it

Work on that, and look at the changes I suggested, and I'll do more reviewing later.

**Comment by dysbulic** [comment]
> I don't know sometimes the search input behaves wierdly becauses it caches the results

Your problem is you are using a `Ref`. Changes to a `Ref` don't cause a rerender of the page, so your updates are going to be erratic.

You should be using `setState`.

File: packages/web/components/MegaMenu.tsx (hunk 1 above)
Comment (Udit-takkar) [reply packages/web/components/MegaMenu.tsx:375]: Ok

File: packages/web/graphql/getPlayers.ts (hunk 4 above)
Comment (Udit-takkar) [reply packages/web/graphql/getPlayers.ts:202]: Will be raising another PR  found a more better way to implement all this

**Pull request closed without merging**